File: discarrays.jl
module Discarrays
    using Mmap

    export discarray, todiscarray

    """
        discarray(io, mode::String="r", type::DataType=Float64, dims=())
    This function is an interface to a memory-mapped array creation. This
    interface has a header of Int64 values, the first one of them being the 
    number of dimensions present in the array, (ndims), and then its dimensions
    (dims). After the header comes the body: a flattened version of that array.
    WARNING: This interface will not save the data types, so you must inform it
    everytime you read a file.
    """
    function discarray(io, mode::String="r", type::DataType=Float64, dims=();
                       shared::Bool=false, pos::Integer=position(io))
        pos !== position(io) && seek(io, pos)
        if occursin("w", mode)
            _dims = Tuple(Int64(d) for d in dims)
            _ndims = Int64(length(_dims))
            write(io, _ndims, _dims...)
        else
            _ndims = read(io, Int64)
            _dims = Tuple(read(io, Int64) for i in 1:_ndims)
        end
        A = Mmap.mmap(io, Array{type, _ndims}, _dims)
        return(A)
    end

    function discarray(array::AbstractArray,
                       mode::String="r", type::DataType=Float64, dims=();
                       shared::Bool=false, pos::Integer=position(io))
        return(array)
    end

    function discarray(filename::String, mode::String="r",
                       type::DataType=Float64, dims=();
                       shared::Bool=false, pos::Integer=0, closeio::Bool=true)
        occursin("w", mode) && @assert dims !== ()
        io = open(filename, mode)
        A = discarray(io, mode, type, dims; shared=shared)
        closeio && close(io)
        return(A)
    end


    """
        todiscarray(filename::String, A::AbstractArray)
    This function takes an array, A, and creates a memory map with read and
    write permissions at a file. This file contains A elements as content after
    discarray header.
    """
    function todiscarray(io, A::AbstractArray)
        discarray(io, "w+", eltype(A), size(A)) .= A
    end

end

File: discarrays.py
import numpy as np


def discarray(filename, mode='r', dtype=float, shape=None, order='C'):
    '''
    This function is an interface to a memory-mapped array creation. This
    interface has a header of Int64 values, the first one of them being the 
    number of dimensions present in the array, (ndim), and then its dimensions
    (shape). After the header comes the body: a flattened version of that array.
    WARNING: This interface will not save the data types, so you must inform it
    everytime you read a file.
    '''
    file_mode = f'{mode[0]}b{mode[1:]}'
    with open(filename, file_mode) as io:
        if mode.startswith('w'):
            assert shape is not None
            shape = tuple(np.asarray(shape).flatten())
            ndims_shape = np.asarray([len(shape), *shape], dtype=np.int64)
            ndims_shape.tofile(io)
        if mode.startswith('r'):
            ndims = np.fromfile(io, dtype=np.int64, count=1)[0]
            shape = tuple(np.fromfile(io, dtype=np.int64, count=ndims))
        offset = io.tell()
        arr = np.memmap(io, mode=mode, dtype=dtype, shape=shape, 
                        offset=offset, order=order)
    return(arr)
